packages/fixie/src/auth.ts

```diff
@@ -0,0 +1,207 @@
+import readline from 'readline';
+import net from 'net';
+
+const { terminal: term } = terminal;
+
+import { FixieClient } from './client.js';
+
+/** Represents contents of the Fixie CLI config file. */
+export interface FixieConfig {
+  apiUrl?: string;
+  apiKey?: string;
+}
+
+export const FIXIE_API_URL = 'https://api.fixie.ai';
+export const FIXIE_CONFIG_FILE = '~/.config/fixie/config.yaml';
+
+/** Load the client configuration from the given file. */
+export function LoadConfig(configFile: string): FixieConfig {
+  const fullPath = untildify(configFile);
+  if (!fs.existsSync(fullPath)) {
+    return {};
+  }
+  const config: object = yaml.load(fs.readFileSync(fullPath, 'utf8')) as object;
+  // Warn if any fields are present in config that are not supported.
+  const validKeys = ['apiUrl', 'apiKey'];
+  const invalidKeys = Object.keys(config).filter((key) => !validKeys.includes(key));
+  for (const key of invalidKeys) {
+    term('❓ Ignoring invalid key ').yellow(key)(` in ${fullPath}\n`);
+  }
+  return config as FixieConfig;
+}
+
+/** Save the client configuration to the given file. */
+export function SaveConfig(config: FixieConfig, configFile: string) {
+  const fullPath = untildify(configFile);
+  const dirName = path.dirname(fullPath);
+  if (!fs.existsSync(dirName)) {
+    fs.mkdirSync(dirName, { recursive: true });
+  }
+  if (fs.existsSync(fullPath)) {
+    // Merge the new config with the existing config, so we don't
+    // overwrite any fields that are not specified.
+    const currentConfig = yaml.load(fs.readFileSync(fullPath, 'utf8')) as object;
+    const mergedConfig = { ...currentConfig, ...config };
+    fs.writeFileSync(fullPath, yaml.dump(mergedConfig));
+  } else {
+    fs.writeFileSync(fullPath, yaml.dump(config));
+  }
+}
+
+/** Returns an authenticated FixieClient, or null if the user is not authenticated. */
+export async function Authenticate({
+  apiUrl,
+  configFile,
+}: {
+  apiUrl?: string;
+  configFile?: string;
+}): Promise<FixieClient | null> {
+  // The precedence for selecting the API URL and key is:
+  //   1. apiUrl argument to this function. (The key cannot be passed as an argument.)
+  //   2. FIXIE_API_URL and FIXIE_API_KEY environment variables.
+  //   3. apiUrl and apiKey fields in the config file.
+  //   4. Fallback value for apiUrl (constant defined above).
+  const config = LoadConfig(configFile ?? FIXIE_CONFIG_FILE);
+  const useApiUrl = apiUrl ?? process.env.FIXIE_API_URL ?? config.apiUrl ?? FIXIE_API_URL;
+  const useApiKey = process.env.FIXIE_API_KEY ?? config.apiKey;
+  if (!useApiKey) {
+    // No key available. Need to punt.
+    return null;
+  }
+  const client = FixieClient.Create(useApiUrl, useApiKey);
+  const userInfo = await client.userInfo();
+  if (userInfo.is_anonymous) {
+    return null;
+  }
+  return client;
+}
+
+/** Returns an authenticated FixieClient, starting an OAuth flow to authenticate the user if necessary. */
+export async function AuthenticateOrLogin({
```

Comment:
```suggestion
export async function AuthenticateOrLogIn({
```

VERY IMPORTANT TOTAL CR BLOCKER

"log in" is a verb; "login" is a noun.